src/models/logits_processor.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include <cstdint>
+#include <cstdio>
+#include <fstream>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <sys/types.h>
+
+#if USE_GUIDANCE
+#include "llguidance.h"
+#endif
+
+#include "logits_processor.h"
+
+namespace Generators {
+
+#if USE_GUIDANCE
+GuidanceLogitsProcessor::GuidanceLogitsProcessor(int vocab_size, uint32_t eos_token,
+                                                 const std::string& guidance_type, const std::string& guidance_data,
+                                                 std::shared_ptr<Tokenizer> tokenizer, const std::string& tokenizer_path)
+    : vocab_size_(vocab_size), eos_token_(eos_token), tokenizer_(std::move(tokenizer)) {
+  if (guidance_type.empty() || guidance_data.empty()) {
+    throw std::runtime_error("Guidance type and data must be provided");
+  }
+
+  if (guidance_type != "json_schema" && guidance_type != "regex" && guidance_type != "grammar") {
+    throw std::runtime_error("Unsupported guidance type: " + guidance_type);
+  }
+
+  auto tokenize_fn = (LlgTokenizeFn) + [](const void* user_data, const uint8_t* bytes,
+                                          size_t bytes_len, uint32_t* output_tokens, size_t output_tokens_len) -> unsigned long {
+    const TokenizeData* tokenize_data = reinterpret_cast<const TokenizeData*>(user_data);
+    auto output_ids = tokenize_partial(reinterpret_cast<const Tokenizer*>(tokenize_data->tokenizer), tokenize_data->prefix_len, bytes, bytes_len);
+    size_t output_size = std::min(output_tokens_len, output_ids.size());
+    for (size_t i = 0; i < output_size; i++) {
+      output_tokens[i] = output_ids[i];
+    }
+    return static_cast<unsigned long>(output_ids.size());
+  };
+
+  // TODO reuse the tokenizer between constraints
+  fs::path tokenizer_path_fs(tokenizer_path);
+  fs::path json_path(tokenizer_path_fs / kDefaultVocabFile);
+  std::ifstream json_file(json_path.string());
+  std::stringstream json_buffer;
+  json_buffer << json_file.rdbuf();
+  std::string json_data = json_buffer.str();
+  auto prefix_len = tokenizer_->Encode(kTokenizePrefixStr).size();
+  tokenize_data_ = {tokenizer_.get(), prefix_len};
+  LlgTokenizerInit tokenizer_init = {
+      static_cast<uint32_t>(vocab_size_),
+      eos_token,
+      nullptr,
```

Comment:
Add comments to describe the arguments.

Reply:
Maybe we could add arg names, leave more details in llguidance.h?

Reply:
Arg names should suffice, thanks.